### ISSUE TITLE
Reciprocal square roots

### ISSUE THREAD

**Issue by NAThompson** [boostorg/multiprecision#239]
Is there a function to compute 1/sqrt(x) in multiprecision?

The Newton’s method for the reciprocal square root does not require division.

**Comment by ckormanyos** [comment]
John might correct me, but if I got it right,...

> Is there a function to compute 1/sqrt(x) in multiprecision?

Thank you for this astute observation.
Nonetheless, i fear the answer is, no.

> The Newton’s method for the reciprocal square root does not require division.

`cpp_dec_float` uses a method called _coutled_ Newton iteration which computes sqrt(x) directly without requiring the inverse operation. Whereas `cpp_bin_float` uses bit shift and bit test on the underlying integral representation.

There is probably optimization rool all around.

Kind regards, Chris

**Comment by pabristow** [comment]
Coupled Newton iteration?  ;-)  (I had to consult the header file to find this! Even dr Google couldn't guess)

**Comment by NAThompson** [comment]
Just found out it's already a canned function [in MPFR](https://pdfs.semanticscholar.org/7767/d043abdd223e646a4a999c74b47e22b57902.pdf).

**Comment by NAThompson** [comment]
Is there a way to detect an MPFR type in an `if constexpr`? I have the following useless hack as a workaround:

```
template<typename Real>
inline Real rsqrt(Real const & x) {
    if constexpr (std::is_same_v<Real, float> || std::is_same_v<Real, double> ||  std::is_same_v<Real, long double>) {
        return 1/sqrt(x);
    }
    if constexpr (std::is_same_v<Real, boost::multiprecision::float128>) {
        return 1/sqrtq(static_cast<__float128>(x.backend().value()));
    }
    else if constexpr (sizeof(Real) > 16) {
        Real res;
        mpfr_rec_sqrt (res.backend().data(), x.backend().data(), GMP_RNDN);
        return res;
    }
}
```

**Comment by NAThompson** [comment]
Fixed in #240

**Comment by jzmaddock** [comment]
At the expense of a little bit of verbosity, I would use good old fashioned overloading for this:

```
template <unsigned Digits10, boost::multiprecision::mpfr_allocation_type AllocateType, boost::multiprecision::expression_template_option ExpressionTemplates>
boost::multiprecision::number<boost::multiprecision::mpfr_float_backend<Digits10, AllocateType>, ExpressionTemplates> rsqrt(const boost::multiprecision::number<boost::multiprecision::mpfr_float_backend<Digits10, AllocateType>, ExpressionTemplates>& arg);

boost::multiprecision::float128 rsqrt(boost::multiprecision::float128 arg);

float rsqrt(float);
double rsqrt(double)
long double rsqrt(long double);
// etc
```

**Comment by NAThompson** [comment]
Fixed in #240
